{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building a genome index for use with loompy and cytograph\n",
    "\n",
    "WARNING: these instruction will create a genome index that is *only* suitable for use with 10x Chromium data. We create a composite index, with separate sequence fragments representing unspliced and spliced transcripts. Unspliced fragments will be generated from genomic sequence upstream of every poly-A stretch located inside of a gene locus. Spliced fragments will be generated from transcriptomic sequence upstream of every polyadenylation site in every known spliced transcript. This reflects the locations of reads generated by Chromium, but may not be suitable for other methods.\n",
    "\n",
    "The instructions below show how to build a genome index for the human genome. For other species, the code below will have to be modified to accomodate different sources and formats of metadata for those species. Instructions for mouse are available in the loompy/kallisto subdirectory.\n",
    "\n",
    "### Install prerequisites\n",
    "\n",
    "Install the [BioPython](https://biopython.org/wiki/Packages) package in your python distribution:\n",
    "\n",
    "```\n",
    "pip install biopython  # Method recommended by the BioPython people\n",
    "conda install -c conda-forge biopython  # Alternative, for Anaconda\n",
    "```\n",
    "\n",
    "Install [gawk](https://www.gnu.org/software/gawk/) (on Linux, this is typically already installed), [bedtools](https://bedtools.readthedocs.io/en/latest/), and [kallisto](https://pachterlab.github.io/kallisto/) (instructions below are for macOS using [Brew](https://brew.s)):\n",
    "```\n",
    "brew install gawk\n",
    "brew install bedtools\n",
    "brew install kallisto\n",
    "```\n",
    "\n",
    "### Download genome data\n",
    "\n",
    "#### Download transcript sequences and metadata from [Gencode](https://www.gencodegenes.org/human/)\n",
    "\n",
    "Get the \"Genome sequence, primary assembly (GRCh38)” file named [GRCh38.primary_assembly.genome.fa.gz](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/GRCh38.primary_assembly.genome.fa.gz)\n",
    "\n",
    "Download the transcript sequences as a fasta file, replacing XX with the version number you want [gencode.vXX.transcripts.fa.gz](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_XX/gencode.vXX.transcripts.fa.gz) . XX = 31 and 38 have been tested.\n",
    "\n",
    "Download the comprehensive PRI gene annotation on the primary assembly GTF file, replacing XX with the version number you want [gencode.vXX.primary_assembly.annotation.gtf](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_XX/gencode.vXX.primary_assembly.annotation.gtf.gz) . XX = 31 and 38 have been tested.\n",
    "\n",
    "#### Download additional gene metadata from [HGNC](https://www.genenames.org/download/statistics-and-files/)\n",
    "\n",
    "Get the “Complete HGNC dataset” as TXT, a file named [hgnc_complete_set.txt](ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/json/hgnc_complete_set.txt)\n",
    "\n",
    "#### Download 10x Chromium barcode whitelists\n",
    "\n",
    "Chromium cell barcodes are generated from a set of known sequences, which differ by version of the Chromium kits. Get three files from the [cellranger GitHub repository](https://github.com/10XGenomics/cellranger/tree/master/lib/python/cellranger/barcodes) and rename them like this:\n",
    "\n",
    "```\n",
    "3M-february-2018.txt.gz -> (unzip) -> 10xv3_whitelist.txt\n",
    "737K-august-2016.txt -> 10xv2_whitelist.txt\n",
    "737K-april-2014_rc.txt -> 10xv1_whitelist.txt\n",
    "```\n",
    "\n",
    "#### Download the human transcription factor motif database\n",
    "\n",
    "Get the [human_tfs_consensus.tab](https://storage.googleapis.com/linnarsson-lab-www-blobs/human_tfs_consensus.tab) file from our Google Cloud. These metadata will be used to determine which genes are transcription factors, and to assign them to families.\n",
    "\n",
    "### Preprocess the genome data\n",
    "\n",
    "Unpack gzip:ed files:\n",
    "\n",
    "```\n",
    "gunzip -c gencode.*.primary_assembly.annotation.gtf.gz > gencode.primary_assembly.annotation.gtf\n",
    "gunzip -c gencode.*.transcripts.fa.gz > gencode.transcripts.fa\n",
    "gunzip GRCh38.primary_assembly.genome.fa.gz\n",
    "```\n",
    "Create a BED file with all the genes:\n",
    "\n",
    "```\n",
    "cat gencode.primary_assembly.annotation.gtf |  gawk 'OFS=\"\\t\" {if ($3==\"gene\") {print $1,$4-1,$5,$10,0,$7}}' | tr -d '\";' > gencode.primary_assembly.annotation.bed\n",
    "```\n",
    "\n",
    "Create a fasta file of pre-mRNA (exons + introns) transcripts:\n",
    "\n",
    "```\n",
    "bedtools sort -i gencode.primary_assembly.annotation.bed > gencode.primary_assembly.annotation.sorted.bed\n",
    "\n",
    "bedtools merge -i gencode.primary_assembly.annotation.sorted.bed -s -c 4 -o collapse > gencode.primary_assembly.annotation.merged.bed\n",
    "\n",
    "bedtools getfasta -name -fo gencode.unspliced.fa -fi GRCh38.primary_assembly.genome.fa -bed gencode.primary_assembly.annotation.sorted.bed\n",
    "```\n",
    "\n",
    "### Create the index manifest file, and an index directory\n",
    "\n",
    "A *manifest* file is necessary to tell cytograph how to find the relevant files in the index. Create a file `manifest.json` with the following content:\n",
    "\n",
    "```\n",
    "{\n",
    "    \"species\": \"Homo sapiens\",\n",
    "    \"index_file\": \"gencode.fragments.idx\",\n",
    "    \"gene_metadata_file\": \"gencode.metadata.tab\",\n",
    "    \"gene_metadata_key\": \"AccessionVersion\",\n",
    "    \"fragments_to_genes_file\": \"fragments2genes.txt\",\n",
    "    \"layers\": {\n",
    "        \"unspliced\": \"unspliced_fragments.txt\",\n",
    "        \"spliced\": \"spliced_fragments.txt\"\n",
    "    }\n",
    "}\n",
    "```\n",
    "\n",
    "The `gene_metadata_key` indicated which column in the metadata file (`gencode.metadata.tab`) contains the gene IDs (as used in `fragments2genes.txt`).\n",
    "\n",
    "Finally, organize your files in a directory and subdirectory as follows:\n",
    "\n",
    "```\n",
    "10xv1_whitelist.txt\n",
    "10xv2_whitelist.txt\n",
    "10xv3_whitelist.txt\n",
    "inputs/\n",
    "    GRCh38.primary_assembly.genome.fa\n",
    "    gencode.unspliced.fa\n",
    "    gencode.primary_assembly.annotation.bed\n",
    "    gencode.primary_assembly.annotation.gtf\n",
    "    gencode.primary_assembly.annotation.sorted.bed\n",
    "    gencode.transcripts.fa\n",
    "    hgnc_complete_set.txt\n",
    "    human_tfs_consensus.tab\n",
    "```\n",
    "\n",
    "The `inputs` directory will only be used to generate the index, and can be discarded afterwards.\n",
    "\n",
    "\n",
    "### Run the following scripts\n",
    "\n",
    "Before starting, set `d` to the full path of your index directory, and `extent` to the desired window size (windows are the sequences upstream of a polyadenylation site or a genomic polyA/T sequence that are used to build the index)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from typing import *\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "d = \"/Users/stelin/kallisto_GRCh38/human_GRCh38_gencode/\"\n",
    "extent = 600  # how many bases away from polya to include\n",
    "min_len = 90  # how many non-repeat bases required to make a transcript"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now run each of the following code blocks:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a gene metadata table from hgnc and gencode gtf\n",
    "# First, load and index the HGNC annotations from ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt\n",
    "hgnc = {}\n",
    "with open(d + \"inputs/hgnc_complete_set.txt\") as f:\n",
    "\thgnc_headers = f.readline()[:-1].split(\"\\t\")\n",
    "\tfor line in f:\n",
    "\t\titems = line[:-1].split(\"\\t\")\n",
    "\t\thgnc[items[0]] = items\n",
    "\n",
    "# Load human consensus TFs\n",
    "tfs = {}\n",
    "with open(d + \"inputs/human_tfs_consensus.tab\") as f:\n",
    "\ttfs_headers = f.readline()[:-1].split(\"\\t\")\n",
    "\tfor line in f:\n",
    "\t\titems = line[:-1].split(\"\\t\")\n",
    "\t\ttfs[items[0]] = items\n",
    "\n",
    "# Next load the gencode GTF and create a genome annotation tsv\n",
    "with open(d + \"gencode.metadata.tab\", \"w\") as fout:\n",
    "\tfout.write(\"\\t\".join([\n",
    "\t\t\"Accession\",\n",
    "\t\t\"AccessionVersion\",\n",
    "\t\t\"Gene\",\n",
    "\t\t\"FullName\",\n",
    "\t\t\"GeneType\",\n",
    "\t\t\"HgncID\",\n",
    "\t\t\"Chromosome\",\n",
    "\t\t\"Strand\",\n",
    "\t\t\"ChromosomeStart\",\n",
    "\t\t\"ChromosomeEnd\",\n",
    "\t\t\"LocusGroup\",\n",
    "\t\t\"LocusType\",\n",
    "\t\t\"Location\",\n",
    "\t\t\"LocationSortable\",\n",
    "\t\t\"Aliases\",\n",
    "\t\t\"VegaID\",\n",
    "\t\t\"UcscID\",\n",
    "\t\t\"RefseqID\",\n",
    "\t\t\"CcdsID\",\n",
    "\t\t\"UniprotID\",\n",
    "\t\t\"PubmedID\",\n",
    "\t\t\"MgdID\",\n",
    "\t\t\"RgdID\",\n",
    "\t\t\"CosmicID\",\n",
    "\t\t\"OmimID\",\n",
    "\t\t\"MirBaseID\",\n",
    "\t\t\"IsTF\",\n",
    "\t\t\"DnaBindingDomain\"\n",
    "\t]))\n",
    "\tfout.write(\"\\n\")\n",
    "\twith open(d + \"inputs/gencode.primary_assembly.annotation.gtf\") as f:\n",
    "\t\tfor line in f:\n",
    "\t\t\tif line.startswith(\"##\"):\n",
    "\t\t\t\tcontinue\n",
    "\t\t\titems = line[:-1].split(\"\\t\")\n",
    "\t\t\tif items[2] != \"gene\":\n",
    "\t\t\t\tcontinue\n",
    "\t\t\textra = {x.strip().split(\" \")[0]: x.strip().split(\" \")[1].strip('\"') for x in items[8].split(\";\")[:-1]}\n",
    "\t\t\tif \"hgnc_id\" in extra and extra[\"hgnc_id\"] in hgnc:\n",
    "\t\t\t\tacc = extra[\"hgnc_id\"]\n",
    "\t\t\t\tensemblID = extra[\"gene_id\"].split(\".\")[0]\n",
    "\t\t\t\tfout.write(\"\\t\".join([\n",
    "\t\t\t\t\tensemblID,\n",
    "\t\t\t\t\textra[\"gene_id\"],\n",
    "\t\t\t\t\thgnc[acc][1],  # gene symbol\n",
    "\t\t\t\t\thgnc[acc][2],  # full name\n",
    "\t\t\t\t\textra[\"gene_type\"],  # gene type from gencode\n",
    "\t\t\t\t\tacc,  # hgnc ID\n",
    "\t\t\t\t\titems[0],  # Chromosome\n",
    "\t\t\t\t\titems[3],  # Start\n",
    "\t\t\t\t\titems[4],  # End\n",
    "\t\t\t\t\thgnc[acc][3],  # Locus group\n",
    "\t\t\t\t\thgnc[acc][4],  # Locus type\n",
    "\t\t\t\t\thgnc[acc][6],  # Location\n",
    "\t\t\t\t\thgnc[acc][7],  # Location, sortable\n",
    "\t\t\t\t\thgnc[acc][8],  # Aliases\n",
    "\t\t\t\t\thgnc[acc][20],  # VEGA id\n",
    "\t\t\t\t\thgnc[acc][21],  # UCSC id\n",
    "\t\t\t\t\thgnc[acc][23],  # Refseq id\n",
    "\t\t\t\t\thgnc[acc][24],  # CCDS id\n",
    "\t\t\t\t\thgnc[acc][25],  # Uniprot id\n",
    "\t\t\t\t\thgnc[acc][26],  # Pubmed id\n",
    "\t\t\t\t\thgnc[acc][27],  # MGD id\n",
    "\t\t\t\t\thgnc[acc][28],  # RGD id\n",
    "\t\t\t\t\thgnc[acc][30],  # COSMIC id\n",
    "\t\t\t\t\thgnc[acc][31],  # OMIM id\n",
    "\t\t\t\t\thgnc[acc][32],  # MIRbase id\n",
    "\t\t\t\t\t\"True\" if (ensemblID in tfs and tfs[ensemblID][3] == \"Yes\") else \"False\", # IsTF?\n",
    "\t\t\t\t\ttfs[ensemblID][2] if (ensemblID in tfs and tfs[ensemblID][3] == \"Yes\") else \"\" # DBD\n",
    "\t\t\t\t]))\n",
    "\t\t\telse:\n",
    "\t\t\t\tensemblID = extra[\"gene_id\"].split(\".\")[0]\n",
    "\t\t\t\tfout.write(\"\\t\".join([\n",
    "\t\t\t\t\tensemblID,\n",
    "\t\t\t\t\textra[\"gene_id\"],\n",
    "\t\t\t\t\textra.get(\"gene_name\", \"\"),  # gene symbol\n",
    "\t\t\t\t\textra.get(\"gene_name\", \"\"),  # full name\n",
    "\t\t\t\t\textra[\"gene_type\"],  # gene type from gencode\n",
    "\t\t\t\t\t\"\",  # HGNC id\n",
    "\t\t\t\t\titems[0],  # Chromosome\n",
    "\t\t\t\t\titems[6],  # Strand\n",
    "\t\t\t\t\titems[3],  # Start\n",
    "\t\t\t\t\titems[4],  # End\n",
    "\t\t\t\t\t\"\",  # Locus group\n",
    "\t\t\t\t\t\"\",  # Locus type\n",
    "\t\t\t\t\t\"\",  # Location\n",
    "\t\t\t\t\t\"\",  # Location, sortable\n",
    "\t\t\t\t\t\"\",  # Aliases\n",
    "\t\t\t\t\t\"\",  # VEGA id\n",
    "\t\t\t\t\t\"\",  # UCSC id\n",
    "\t\t\t\t\t\"\",  # Refseq id\n",
    "\t\t\t\t\t\"\",  # CCDS id\n",
    "\t\t\t\t\t\"\",  # Uniprot id\n",
    "\t\t\t\t\t\"\",  # Pubmed id\n",
    "\t\t\t\t\t\"\",  # MGD id\n",
    "\t\t\t\t\t\"\",  # RGD id\n",
    "\t\t\t\t\t\"\",  # COSMIC id\n",
    "\t\t\t\t\t\"\",  # OMIM id\n",
    "\t\t\t\t\t\"\",  # MIRbase id\n",
    "\t\t\t\t\t\"True\" if (ensemblID in tfs and tfs[ensemblID][3] == \"Yes\") else \"False\", # IsTF?\n",
    "\t\t\t\t\ttfs[ensemblID][2] if (ensemblID in tfs and tfs[ensemblID][3] == \"Yes\") else \"\" # DBD\n",
    "\t\t\t\t]))\n",
    "\t\t\tfout.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_polys(seq: SeqRecord, c: str = \"A\", n: int = 15) -> List[Tuple[int, int]]:\n",
    "\tfound = []\n",
    "\tcount = seq[:n].count(c)  # Count occurences in the first k-mer\n",
    "\tif count >= n - 1:  # We have a match\n",
    "\t\tfound.append(0)\n",
    "\tix = 0\n",
    "\twhile ix < len(seq) - n - 1:\n",
    "\t\tif seq[ix] == c:  # Outgoing base\n",
    "\t\t\tcount -= 1\n",
    "\t\tif seq[ix + n] == c:  # Incoming base\n",
    "\t\t\tcount += 1\n",
    "\t\tix += 1\n",
    "\t\tif count >= n - 1:  # We have a match\n",
    "\t\t\tfound.append(ix)\n",
    "\t\n",
    "\tsorted_by_lower_bound = [(f, f + n) for f in found]\n",
    "\t# merge intervals (https://codereview.stackexchange.com/questions/69242/merging-overlapping-intervals)\n",
    "\tmerged = []\n",
    "\tfor higher in sorted_by_lower_bound:\n",
    "\t\tif not merged:\n",
    "\t\t\tmerged.append(higher)\n",
    "\t\telse:\n",
    "\t\t\tlower = merged[-1]\n",
    "\t\t\t# test for intersection between lower and higher:\n",
    "\t\t\t# we know via sorting that lower[0] <= higher[0]\n",
    "\t\t\tif higher[0] <= lower[1]:\n",
    "\t\t\t\tupper_bound = max(lower[1], higher[1])\n",
    "\t\t\t\tmerged[-1] = (lower[0], upper_bound)  # replace by merged interval\n",
    "\t\t\telse:\n",
    "\t\t\t\tmerged.append(higher)\n",
    "\treturn merged\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "polyAs = {}\n",
    "polyTs = {}\n",
    "for fasta in SeqIO.parse(open(d + \"inputs/gencode.unspliced.fa\"),'fasta'):\n",
    "\tgene_id = fasta.id\n",
    "\tintervals = find_polys(fasta.seq, c=\"A\", n=14)\n",
    "\tif len(intervals) > 0:\n",
    "\t\tpolyAs[gene_id] = intervals\n",
    "\t# Collect fragments on the opposite strand, downstream of poly-Ts (not sure if such reads really happen?)\n",
    "\tintervals = find_polys(fasta.seq, c=\"T\", n=14)\n",
    "\tif len(intervals) > 0:\n",
    "\t\tpolyTs[gene_id] = intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr2g = {}\n",
    "with open(d + \"inputs/gencode.primary_assembly.annotation.gtf\") as f:\n",
    "\tfor line in f:\n",
    "\t\tif \"\\ttranscript\\t\" in line:\n",
    "\t\t\titems = line.split(\"; \")\n",
    "\t\t\tchrom, _, _, start, end, _, strand, _, gid = items[0].split(\"\\t\")\n",
    "\t\t\tgene_id = gid.split('\"')[1]\n",
    "\t\t\ttranscript_id = items[1].split('\"')[1]\n",
    "\t\t\tgene_type = items[2].split('\"')[1]\n",
    "\t\t\tgene_name = items[3].split('\"')[1]\n",
    "\t\t\ttr2g[transcript_id] = (chrom, start, end, strand, gene_id, gene_type, gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 0\n",
    "with open(d + \"fragments2genes.txt\", \"w\") as ftr2g:\n",
    "\twith open(d + \"inputs/gencode.fragments.fa\", \"w\") as fout:\n",
    "\t\t# Write the nascent fragments, with one partial transcript per internal poly-A/T site\n",
    "\t\twith open(d + \"unspliced_fragments.txt\", \"w\") as fucapture:\n",
    "\t\t\tfor fasta in SeqIO.parse(open(d + \"inputs/gencode.unspliced.fa\"),'fasta'):  # Note we're in the masked file now\n",
    "\t\t\t\tgene_id = fasta.id\n",
    "\t\t\t\tif gene_id in polyAs:\n",
    "\t\t\t\t\tfor interval in polyAs[gene_id]:\n",
    "\t\t\t\t\t\tseq = str(fasta.seq[max(0, interval[0] - extent):interval[0]])\n",
    "\t\t\t\t\t\t#seq = seq.translate(tr).strip(\"N\")\n",
    "\t\t\t\t\t\tif len(seq) >= min_len:\n",
    "\t\t\t\t\t\t\tcount += 1\n",
    "\t\t\t\t\t\t\ttranscript_id = f\"{gene_id}.A{interval[0]}\"\n",
    "\t\t\t\t\t\t\ttrseq = SeqRecord(Seq(seq), transcript_id, '', '')\n",
    "\t\t\t\t\t\t\tfout.write(trseq.format(\"fasta\"))\n",
    "\t\t\t\t\t\t\tftr2g.write(f\"{transcript_id}\\t{gene_id}\\n\")\n",
    "\t\t\t\t\t\t\tfucapture.write(f\"{transcript_id}\\n\")\n",
    "\t\t\t\tif gene_id in polyTs:\n",
    "\t\t\t\t\tfor interval in polyTs[gene_id]:\n",
    "\t\t\t\t\t\tseq = str(fasta.seq[interval[1]:interval[1] + extent].reverse_complement())\n",
    "\t\t\t\t\t\t#seq = seq.translate(tr).strip(\"N\")\n",
    "\t\t\t\t\t\tif len(seq) >= min_len:\n",
    "\t\t\t\t\t\t\tcount += 1\n",
    "\t\t\t\t\t\t\ttranscript_id = f\"{gene_id}.T{interval[0]}\"\n",
    "\t\t\t\t\t\t\ttrseq = SeqRecord(Seq(seq), transcript_id, '', '')\n",
    "\t\t\t\t\t\t\tfout.write(trseq.format(\"fasta\"))\n",
    "\t\t\t\t\t\t\tftr2g.write(f\"{transcript_id}\\t{gene_id}\\n\")\n",
    "\t\t\t\t\t\t\tfucapture.write(f\"{transcript_id}\\n\")\n",
    "\t\t# Write the mature fragments, covering the 3' end of each mature transcript\n",
    "\t\twith open(d + \"spliced_fragments.txt\", \"w\") as fscapture:\n",
    "\t\t\tfor fasta in SeqIO.parse(open(d + \"inputs/gencode.transcripts.fa\"),'fasta'):  # Note we're in the masked file now\n",
    "\t\t\t\ttranscript_id = fasta.id.split(\"|\")[0]\n",
    "\t\t\t\tgene_id = fasta.id.split(\"|\")[1]\n",
    "\t\t\t\tattrs = tr2g[transcript_id]\n",
    "\t\t\t\tseq = str(fasta.seq[-extent:])\n",
    "\t\t\t\tif len(seq) >= min_len:\n",
    "\t\t\t\t\tcount += 1\n",
    "\t\t\t\t\ttrseq = SeqRecord(Seq(seq), f\"{transcript_id}.{count} gene_id:{attrs[4]} gene_name:{attrs[6]}\", '', '')\n",
    "\t\t\t\t\tfout.write(trseq.format(\"fasta\"))\n",
    "\t\t\t\t\tftr2g.write(f\"{transcript_id}.{count}\\t{attrs[4]}\\n\")\n",
    "\t\t\t\t\tfscapture.write(f\"{transcript_id}.{count}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build the kallisto index\n",
    "\n",
    "Run the following on the command line (it might take half an hour):\n",
    "\n",
    "```\n",
    "kallisto index -i gencode.fragments.idx -k 31 inputs/gencode.fragments.fa\n",
    "```\n",
    "\n",
    "You should now have the following directory structure:\n",
    "\n",
    "```\n",
    "10xv1_whitelist.txt\n",
    "10xv2_whitelist.txt\n",
    "10xv3_whitelist.txt\n",
    "fragments2genes.txt\n",
    "gencode.fragments.idx\n",
    "gencode.metadata.tab\n",
    "manifest.json\n",
    "spliced_fragments.txt\n",
    "unspliced_fragments.txt\n",
    "inputs/\n",
    "```\n",
    "\n",
    "You can now remove the `inputs/` subdirectory.\n",
    "\n",
    "### What just happened?\n",
    "\n",
    "We created a composite kallisto index (`gencode.fragments.idx`), with separate sequence fragments for spliced and unspliced transcripts. The IDs of spliced and unspliced fragments are listed in the `spliced_fragments.txt` and `unspliced_fragments.txt` files, so that we can later count them separately. For example, here's the first few lines of `unspliced_fragments.txt` (the last part of each ID, e.g. `A14056`, indicates that this fragment is located upstream of a poly-A stretch at position 14056):\n",
    "\n",
    "```\n",
    "ENSG00000277400.1.A14056\n",
    "ENSG00000277400.1.A32841\n",
    "ENSG00000277400.1.A35311\n",
    "ENSG00000277400.1.A36796\n",
    "ENSG00000277400.1.A44325\n",
    "ENSG00000277400.1.A45592\n",
    "ENSG00000277400.1.A47356\n",
    "ENSG00000277400.1.A49571\n",
    "ENSG00000277400.1.A53084\n",
    "ENSG00000277400.1.A53231\n",
    "```\n",
    "\n",
    "We also created a mapping from fragments to genes, `fragments2genes.txt` so that we can later pool counts by gene. Here's the first few lines of that file:\n",
    "\n",
    "```\n",
    "ENSG00000277400.1.A14056\tENSG00000277400.1\n",
    "ENSG00000277400.1.A32841\tENSG00000277400.1\n",
    "ENSG00000277400.1.A35311\tENSG00000277400.1\n",
    "ENSG00000277400.1.A36796\tENSG00000277400.1\n",
    "ENSG00000277400.1.A44325\tENSG00000277400.1\n",
    "ENSG00000277400.1.A45592\tENSG00000277400.1\n",
    "ENSG00000277400.1.A47356\tENSG00000277400.1\n",
    "ENSG00000277400.1.A49571\tENSG00000277400.1\n",
    "ENSG00000277400.1.A53084\tENSG00000277400.1\n",
    "ENSG00000277400.1.A53231\tENSG00000277400.1\n",
    "```\n",
    "\n",
    "Finally, we created a consolidated metadata file `gencode.metadata.tab` which collects a lot of useful annotation about each gene:\n",
    "\n",
    "```\n",
    "Accession          # ENSEMBL accession\n",
    "AccessionVersion   # ENSEMBL accession.version\n",
    "Gene               # HGNC official gene symbol\n",
    "FullName           # Gene long name\n",
    "GeneType           # Like 'protein_coding', 'pseudogene', 'snRNA', 'rRNA', etc.\n",
    "HgncID\n",
    "Chromosome         # Like 'chr1', 'chr2', 'chrM', etc.\n",
    "Strand             # '+' or '-'\n",
    "ChromosomeStart    # Integer start position of the gene\n",
    "ChromosomeEnd      # Integer end position of the gene\n",
    "LocusGroup\n",
    "LocusType\n",
    "Location           # Like '1p36.33'\n",
    "LocationSortable   # Like '01p36.33'\n",
    "Aliases\n",
    "VegaID\n",
    "UcscID\n",
    "RefseqID\n",
    "CcdsID\n",
    "UniprotID\n",
    "PubmedID\n",
    "MgdID\n",
    "RgdID\n",
    "CosmicID\n",
    "OmimID\n",
    "MirBaseID\n",
    "IsTF                # 'True' if the gene is a transcription factor\n",
    "DnaBindingDomain    # Like 'HMG/Sox', 'Homeodomain', etc.\n",
    "```\n",
    "\n",
    "Of these, only `Accession` and `Gene` are required by Cytograph.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
